File: 008_image_filtering_using_convolution/CPP/image_convolution.cpp
// Import dependancies
#include <opencv2/opencv.hpp>
#include <iostream>

// Using namespaces to nullify use of c::function(); syntax and std::function(); syntax
using namespace std;
using namespace cv;

int main()
{
    // Read Image
    Mat image = imread("test.jpg");

    // Print Error message if image is null
    if (image.empty()) 
        {
            cout << "Could not read image" << endl;
        }
    
    // Apply identity filter using kernel
    Mat kernel1 = (Mat_<double>(3,3) << 0, 0, 0, 0, 1, 0, 0, 0, 0);
    Mat identity; 
    filter2D(image, identity, -1 , kernel1, Point(-1, -1), 0, 4);
    imshow("Original", image);
    imshow("Identity", identity);
    waitKey();
    imwrite("identity.jpg", identity);
    destroyAllWindows();

    // Blurred using kernel
    // Initialize matrix with all ones
    Mat kernel2 = Mat::ones(5,5, CV_64F);
    // Normalize the elements
    kernel2 = kernel2 / 25;
    Mat img;
    filter2D(image, img, -1 , kernel2, Point(-1, -1), 0, 4);
    imshow("Original", image);
    imshow("Kernel blur", img);
    imwrite("blur_kernel.jpg", img);
    waitKey();
    destroyAllWindows();

    // Blurred using OpenCV C++ blur() function
    Mat img_blur;
    blur(image, img_blur, Size(5,5));
    imshow("Original", image);
    imshow("Blurred", img_blur); 
    imwrite("blur.jpg", img_blur);
    waitKey();
    destroyAllWindows();
    
    // Performing Gaussian Blur
    Mat gaussian_blur;
    GaussianBlur(image, gaussian_blur, Size(5,5), 0);
    imshow("Original", image);
    imshow("Gaussian Blurred", gaussian_blur);
    imwrite("gaussian_blur.jpg", gaussian_blur);
    waitKey();
    destroyAllWindows();

    // Apply Median Blur
    Mat median_blurred;
    medianBlur(image, median_blurred, (5,5));
    imshow("Original", image);
    imshow("Median Blurred", median_blurred);
    imwrite("median_blur.jpg", median_blurred);
    waitKey();
    destroyAllWindows();

    // Apply Sharpening using kernel
    Mat sharp_img;
    Mat kernel3 = (Mat_<double>(3,3) << 0, -1, 0, -1, 5, -1, 0, -1 ,0);
    filter2D(image, sharp_img, -1 , kernel3, Point(-1, -1), 0, BORDER_DEFAULT);
    imshow("Original", image);
    imshow("Sharpenned", sharp_img);
    imwrite("sharp_image.jpg", sharp_img);
    waitKey();
    destroyAllWindows();

    // Apply bilateral filtering
    Mat bilateral_filter;
    bilateralFilter(image, bilateral_filter, 9, 75, 75);
    imshow("Original", image);
    imshow("Bilateral filtering", bilateral_filter);
    imwrite("bilateral_filtering.jpg", bilateral_filter);
    waitKey(0);
    destroyAllWindows();
    return 0;
}

File: 008_image_filtering_using_convolution/Python/image_convolution.py
import cv2
import numpy as np

image = cv2.imread('test.jpg')

# Print error message if image is null
if image is None:
    print('Could not read image')

# Apply identity kernel
kernel1 = np.array([[0, 0, 0],
                   [0, 1, 0],
                   [0, 0, 0]])

identity = cv2.filter2D(src=image, ddepth=-1, kernel=kernel1)

cv2.imshow('Original', image)
cv2.imshow('Identity', identity)
    
cv2.waitKey()
cv2.imwrite('identity.jpg', identity)
cv2.destroyAllWindows()

# Apply blurring kernel
kernel2 = np.ones((5, 5), np.float32) / 25
img = cv2.filter2D(src=image, ddepth=-1, kernel=kernel2)

cv2.imshow('Original', image)
cv2.imshow('Kernel Blur', img)
    
cv2.waitKey()
cv2.imwrite('blur_kernel.jpg', img)
cv2.destroyAllWindows()

# Apply blur using `blur()` function
img_blur = cv2.blur(src=image, ksize=(5,5)) 
cv2.imshow('Original', image)
cv2.imshow('Blurred', img_blur)
    
cv2.waitKey()
cv2.imwrite('blur.jpg', img_blur)
cv2.destroyAllWindows()

# Apply Gaussian blur
gaussian_blur = cv2.GaussianBlur(src=image, ksize=(5,5), sigmaX=0)

cv2.imshow('Original', image)
cv2.imshow('Gaussian Blurred', gaussian_blur)
    
cv2.waitKey()
cv2.imwrite('gaussian_blur.jpg', gaussian_blur)
cv2.destroyAllWindows()

# Apply Median blur
median_blurred = cv2.medianBlur(src=image, ksize=5)

cv2.imshow('Original', image)
cv2.imshow('Median Blurred', median_blurred)
cv2.imwrite('median_blur.jpg', median_blurred)
cv2.waitKey()
cv2.destroyAllWindows()

# Apply sharpening using kernel
kernel3 = np.array([[0, -1, 0],
                   [-1, 5, -1],
                   [0, -1, 0]])
sharp_img = cv2.filter2D(src=image, ddepth=-1, kernel=kernel3)

cv2.imshow('Original', image)
cv2.imshow('Sharpened', sharp_img)
    
cv2.waitKey()
cv2.imwrite('sharp_image.jpg', sharp_img)
cv2.destroyAllWindows()

# Apply Bilateral Filtering
bilateral_filter = cv2.bilateralFilter(src=image, d=9, sigmaColor=75, sigmaSpace=75)

cv2.imshow('Original', image)
cv2.imshow('Bilateral Filtering', bilateral_filter)
cv2.imwrite('bilateral_filtering.jpg', bilateral_filter)
cv2.waitKey(0)

cv2.destroyAllWindows()